core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1608,16 +1629,20 @@ public long newRequestId() {
      * Called once the channel is closed for instance due to a disconnect or a closed socket etc.
      */
     protected final void onChannelClosed(Channel channel) {
-        final Optional<Long> first = pendingHandshakes.entrySet().stream()
-            .filter((entry) -> entry.getValue().channel == channel).map((e) -> e.getKey()).findFirst();
-        if (first.isPresent()) {
-            final Long requestId = first.get();
-            final HandshakeResponseHandler handler = pendingHandshakes.remove(requestId);
-            if (handler != null) {
-                // there might be a race removing this or this method might be called twice concurrently depending on how
-                // the channel is closed ie. due to connection reset or broken pipes
-                handler.handleException(new TransportException("connection reset"));
+        try {
+            final Optional<Long> first = pendingHandshakes.entrySet().stream()
+                .filter((entry) -> entry.getValue().channel == channel).map((e) -> e.getKey()).findFirst();
+            if (first.isPresent()) {
+                final Long requestId = first.get();
+                final HandshakeResponseHandler handler = pendingHandshakes.remove(requestId);
+                if (handler != null) {
+                    // there might be a race removing this or this method might be called twice concurrently depending on how
+                    // the channel is closed ie. due to connection reset or broken pipes
+                    handler.handleException(new TransportException("connection reset"));
+                }
             }
+        } finally {
+            disconnectFromNodeChannel(channel, "channel closed");
```

Comment:
you need this since otherwise we might not notify a the listeners waiting for an response on the connection. We might be able to reduce it's footprint more. I can look but it's already improved and preexsting hence why I don't wanted to touch it more. The call was hidden by other calls that's why you don't see it